[WIP][MIG] Migration of project_scrum to 10.0

daramousk opened the pull request:
Takes things over from #1

pedrobaeza commented:
There are .bzr files inside i18n

hbrunn commented on project_scrum/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+Usage
+=====
+This module implements all concepts defined by the scrum project management methodology for IT companies: * Project with sprints, product owner, scrum master * Sprints with reviews, daily meetings, feedbacks * Product backlog * Sprint backlog
```
make a bullet list out of this

hbrunn commented on project_scrum/README.rst:
```diff
@@ -0,0 +1,52 @@
+
+* Mohamed Habib Challouf <medhabib@tenovar.com>
+* Samir Guesmi <samir@tenovar.com>
+* George Daramouskas <gdaramouskas@therp.nl>
```
probably we should add the original authors here. @vertelab, can you PR what you want to have here (and in the manifest) so that we don't lose your credits?

hbrunn commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,96 @@
+    project_id = fields.Many2one(
+        comodel_name='project.project',
+        string='Project',
+        ondelete='set null',
```
I think this field should be required and `ondelete='cascade'`

hbrunn commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,96 @@
+        required=True,
+        default=lambda self: self.env.user
+    )
+    question_yesterday = fields.Text(
```
all of those should be html fields

hbrunn commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,96 @@
+        related='project_id.analytic_account_id.company_id',
+    )
+
+    def _compute_meeting_name(self):
```
better stick this in `name_get` and remove the `name` field entirely

hbrunn commented on project_scrum/models/project_scrum_sprint.py:
```diff
@@ -0,0 +1,191 @@
+            ('project_id', '=', project_id)
+        ])
+        if sprint:
+            return sprint[0]
```
set `limit=1` above instead. actually, you can just return the search with the limit, it will be an emtpy recordset if the search is empty

hbrunn commented on project_scrum/models/project_scrum_sprint.py:
```diff
@@ -0,0 +1,191 @@
+    project_id = fields.Many2one(
+        comodel_name='project.project',
+        string='Project',
+        ondelete='set null',
```
this also should be cascade and required

hbrunn commented on project_scrum/models/project_scrum_sprint.py:
```diff
@@ -0,0 +1,191 @@
+    @api.onchange('date_start')
+    def onchange_date_start(self):
+        if self.date_start:
+            if self.project_id:
```
combine those two

hbrunn commented on project_scrum/models/project_scrum_test.py:
```diff
@@ -0,0 +1,63 @@
+    project_id = fields.Many2one(
+        comodel_name='project.project',
+        string='Project',
+        ondelete='set null',
```
required + cascade

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,164 @@
+        if isinstance(context.get('default_project_id'), basestring):
+            project_name = context['default_project_id']
+            project_ids = self.env['project.project'].with_context(
+                context).name_search(name=project_name)
```
pass `operator` too, the default `ilike` won't yield what we want, or only fall back to that when `=` or `=ilike` don't yield results

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,156 @@
+
+    @api.model
+    def create(self, vals):
+        vals['reference'] = self.env['ir.sequence'].next_by_code(
```
check if this is created within a project that has `use_scrum` activated

hbrunn replied:
or (better) remove entirely and depend on https://github.com/OCA/project/tree/10.0/project_task_code

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,156 @@
+            ('not_set', 'Not Set'),
+        ]
+
+    user_id = fields.Many2one(
```
this and the date fields look like redefinitions from the project module. Remove if that's true

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,156 @@
+    )
+    date_end = fields.Datetime(string='Ending Date')
+    use_scrum = fields.Boolean(related='project_id.use_scrum')
+    description = fields.Html('Description')
```
this too

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,156 @@
+    def write(self, vals):
+        if vals.get('stage_id') == self.env.ref(
+                'project.project_stage_data_2').id:
+            vals['date_end'] = fields.date.context_today()
```
this looks more like something we want in https://github.com/OCA/project/blob/10.0/project_stage_state and have this module depend on it

hbrunn reviewed (changes requested):
what I miss in this module are decidated security groups for the different roles in a scrum project and the UI reactiving to that.
And the access rules for user stories et al should mirror the access rules for tasks depending on the visiblity settings of the project as in https://github.com/OCA/OCB/blob/10.0/addons/project/security/project_security.xml

daramousk commented:
> what I miss in this module are decidated security groups for the different roles in a scrum project and the UI reactiving to that.

@hbrunn Can you be more specific here? I have added the three roles that exist on scrum but I am unsure as to which views/fields should they be able to view.

hbrunn commented:
you're more into scrum than me, so I think your opinion is as good as mine. I assume owner and/or the master should be able to create and change user stories, change states there and such, and scrum users should only be able to change whatever it is they change.
Some of the buttons probably are only relevant for certain roles too I guess

Commit:
[PORT] Forward porting project_scrum to 10.0

hbrunn commented on project_scrum/views/project_scrum_view.xml:
```diff
@@ -0,0 +1,1138 @@
+            <field name="model">project.task</field>
+            <field name="inherit_id" ref="project.view_task_form2" />
+            <field name="priority">99</field>
+            <field name="arch" type="xml">
```
I think we need a separation of scrum tasks and non-scrum tasks. So better make this a primary view, and add a menu that forces this view. Do the same with actions that open tasks of a scrum project.

Alternatively, change this view such that it doesn't change anything for non-scrum tasks.

daramousk commented:
@hbrunn Some of the security guidelines were already implemented and I have made some changes as well, got some help from @lfreeke. Tests successful as well.

lfreeke commented:
@daramousk Create a separate task view for scrum projects. In case people are using both scrum and non scrum projects. You now see the scrum fields in the task tree view. I think it clearer for the user when it is separated.

hbrunn commented:
it might make sense to do the same for projects themselves too. Then you can do nifty stuff like also passing a different view in the tasks button on the form, and have a different kanban/tree view vor them too.

Implement using a primary view override, this way you don't have to recreate everything from scratch but basically don't have to change much from what you have now

hbrunn commented:
PS: also think of nice stuff like setting `default_use_scrum` and the like in the context of the actions you add

daramousk commented:
@lfreeke @hbrunn This functionality already exists though, you can see the `Use Scrum?` boolean which you can toggle both on `project.project` and `project.task`. It shows only the relevant views. Is this not working properly? Or should this functionality refactored?

hbrunn commented:
there are several problems with the current approach:
- the treeview for tasks shows a lot of unnecessary column for non-scrum-tasks
- hidden columns (and fields) still cause access errors for people who have project permissions, but no scrum ones
- use_scrum for a task should inherit this from the project, a standalone scrum task doesn't make sense

And of course if we have custom views for the scrum records we can do fancy stuff there that we couldn't properly do when having to keep the dual use form.

So just add a projects menu under the scrum item, and in the standard form only add the `use scrum` to the standard form so that we can create scrum projects either way. Under the scope of the scrum/projects menu, that should be the default.

daramousk commented:
@NL66278 A review here as well if possible please.

NL66278 commented on project_scrum/demo/project_scrum_demo.xml:
```diff
@@ -0,0 +1,194 @@
+    <record id="project_sprint_1" model="project.scrum.sprint">
+        <field name="name">Sprint 1</field>
+        <field name="project_id" ref="project_project_1"/>
+        <field name="date_start" eval="(DateTime.today() - relativedelta(weeks=1)).strftime('%Y-%m-%d')"/>
```
With very long lines, put each attribute on a separate line. Now for instance I need to scroll horizontally to see what is in the eval.

NL66278 commented on project_scrum/models/project_project.py:
```diff
@@ -0,0 +1,76 @@
+    user_story_ids = fields.One2many(
+        comodel_name="project.scrum.us",
+        inverse_name="project_id",
+        string="User Sories",
```
type User Stories is missing the t in Stories

NL66278 commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,99 @@
+        string='Sprint',
+    )
+    date_meeting = fields.Date(
+        string='Date',
```
There are a lot of places in Odo where I think, why a datetime if only a date is needed. Here I feel the opposite. A meeting starts usually at a specific time, so why not a datetime?

NL66278 commented on project_scrum/views/project_scrum_test.xml:
```diff
@@ -0,0 +1,171 @@
+                    </templates>
+                </kanban>
+            </field>
+        </record>        <record id="view_ps_tc_tree" model="ir.ui.view">
```
Split this line

NL66278 commented on project_scrum/security/res_groups.xml:
```diff
@@ -0,0 +1,15 @@
+<odoo>
```
Missing: <?xml version="1.0" encoding="utf-8"?>

(Also missing from other views, please add them to all xml documents).

daramousk replied:
@NL66278 I do not understand this comment. What exactly is missing from this file?

NL66278 replied:
Sorry forgot to paste: <?xml version="1.0" encoding="utf-8"?>

NL66278 replied:
No, it is an error in github comments. retry:
 ```<?xml version="1.0" encoding="utf-8"?>```

NL66278 commented on project_scrum/views/project_scrum_sprint.xml:
```diff
@@ -0,0 +1,262 @@
+                                    <tree>
+                                        <field name="name" />
+                                        <field name="actor_ids" />
+                                        <field name="sprint_ids" />
```
You can leave out the fields sprint_ids and project_id here, because they just repeat in each line information that is already in the main part of the form.

NL66278 commented on project_scrum/__manifest__.py:
```diff
@@ -0,0 +1,41 @@
+        'project_stage_state',
+        'mail',
+        'hr_timesheet',
+        'web_kanban',
```
You do not need these dependencies on the web_* modules

NL66278 commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,172 @@
+        return super(ProjectScrumUs, self).create(vals)
+
+    @api.model
+    def _get_moscow_field(self):
```
Return the values from project.task. No you define the moscow selection in two separate places.

NL66278 commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,99 @@
+        string='Date',
+        required=True,
+    )
+    user_id_meeting = fields.Many2one(
```
So these are "meetings" where only one person can be present?

daramousk replied:
@NL66278 Going through the rest of the code and the templates I see that this is supposed to be the organizer of the meeting. I might be wrong though, you might want to check the mail_template.xml. And if this field is turned to many2many how will the template change?

NL66278 reviewed (commented):
In general the migration seems to be working well.

I have one major issue, and that is the dependency on hr_timesheet. This means installing project_scrum is impossible without installing accounting. Actually hr_timesheet is only used to count progress in hours for sprints.  Maybe that functionality can be split of in a separate module that depends on project_scrum and hr_timesheet and autoinstalls when both are installed.

For the rest some nitpickings and the need to remove the dpendency on the web_* modules.

NL66278 commented on project_scrum/models/project_scrum_meeting.py:
```diff
@@ -0,0 +1,99 @@
+        string='Date and Time of the meeting',
+        required=True,
+    )
+    user_id_meeting = fields.Many2one(
```
You might change the string to "Meeting organiser" then you can keep it a Many2one, but it will be more clear why you can only select one person.

daramousk commented:
@NL66278 Can you tell me which functionality here requires `hr_timesheet`? I have removed it and everything seems to work fine, I have changed the name as well.

NL66278 commented:
@daramousk You are dependent on hr_timesheet, because you use the effective_hours added by hr_timesheet to project.task. Your present code crashes when trying to open sprints.

daramousk commented:
@NL66278 The functionality that has to be removed is quite big though, the `effective_hours` is used in a couple of places among them. Also, the calculation of `progress` needs it. 

The `effective_hours` is needed to calculate the progress of the `sprint` and the progress is essential to be known as per the sprint guide https://www.scrum.org/resources/what-is-a-sprint-in-scrum

NL66278 commented:
@daramousk I realise that effective_hours is the base of a lot of functionality. My proposal was not simply to remove it, but to put it in a separate module that is autoinstall when both hr_timesheet and project_scrum are installed.
The real problem, which we can do nothing about for the moment, is that hr_timesheet should never have been dependent on accounting.
On the other hand, we have been perfectly happy using scrum without doing anything with checking progress based on hours burned. We can also gauge progress by wether the issues assigned to a sprint are actually resolved.

daramousk commented:
@NL66278 Removed this functionality

NL66278 reviewed (approved):
:+1: LGTM . Also did some user testing and worked well.

I suggested to move the parts that are dependent on hr_timesheet (and therefore accounting) to a separate module. That might go in another PR.

daramousk commented:
@hbrunn Possibly this is ready for merge?

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,167 @@
+        return self.env['project.task']._get_moscow_field()
+
+    @api.model
+    def _get_value_field(self):
```
any reason not to do the same as above here?

hbrunn replied:
and below

hbrunn commented on project_scrum/models/project_scrum_actors.py:
```diff
@@ -0,0 +1,11 @@
+    _name = 'project.scrum.actors'
+    _description = 'Actors in user stories'
+
+    name = fields.Char(size=60)
```
this probably wants to be required. Why the size restriction?

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,167 @@
+        ondelete='set null',
+        index=True,
+        track_visibility='onchange',
+        change_default=True,
```
why isn't this required?

hbrunn commented on project_scrum/security/project_security.xml:
```diff
@@ -0,0 +1,81 @@
+            </field>
+        </record>
+
+        <record model="ir.rule" id="ps_actors_rule">
```
what dies this do?

hbrunn commented on project_scrum/security/project_security.xml:
```diff
@@ -0,0 +1,81 @@
+        <record model="ir.rule" id="ps_us_rule_visibility">
+            <field name="name">User Story Visibility</field>
+            <field name="model_id" ref="model_project_scrum_us" />
+            <field name="domain_force">
```
shouldn't we have something similar to this for sprints, meetings, tests?

daramousk replied:
I am not sure about this one. In any case this is supposed to be a migration, should we go and change the security? Possibly there where some requirements initially that we are unaware of?

hbrunn replied:
it's not just a migration, this is the first version that goes into the OCA.

Consider the way we're going to use this, having different customers log in on our instance and scrumming with us. Do we really want them to be able to see other customers' sprints etc?

daramousk replied:
@hbrunn Added the extra security. Now a user needs to be invited in order to see sprints, meetings, tests, user stories for projects that are by invitation only.

hbrunn commented on project_scrum/tests/test_project_scrum.py:
```diff
@@ -0,0 +1,213 @@
+_logger = logging.getLogger(__name__)
+
+
+class TestProjectScrum(TransactionCase):
```
without sensible assertions, most of those tests are pointless, that's just fake code coverage. In the create test, you can for example assert that it's the correct display name for meetings, correct counts for user stories, etc.
Generally, if you don't assert correct functioning of the code you call in a test, you can just as well leave out the test, because you just run code, don't test it

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,131 @@
+    risk = fields.Selection('_get_risk_field')
+    kano = fields.Selection('_get_kano_field')
+    color = fields.Integer(related='project_id.color')
+    reference = fields.Char(
```
is this just some arbitrary code for tasks? If so, replace this by a dependency to https://github.com/OCA/project/tree/10.0/project_task_code

daramousk commented:
@hbrunn Implemented all the changes

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,144 @@
+        return self.env['project.task']._get_risk_field()
+
+    @api.model
+    def _get_kano_field(self):
```
get those from `project.task` too

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,144 @@
+    _group_by_full = {
+        'sprint_ids': _read_group_sprint_id,
+        }
+    name = fields.Char()
```
what's this good for?

hbrunn commented on project_scrum/models/project_scrum_us.py:
```diff
@@ -0,0 +1,144 @@
+            [('project_id', '=', project_id)], order='sequence').name_get()
+        return sprints, None
+
+    _group_by_full = {
```
this is a relict from 8.0: https://github.com/OCA/OCB/blob/8.0/openerp/models.py#L1848

You probably need https://github.com/OCA/OCB/blob/10.0/odoo/models.py#L1671 here

hbrunn commented on project_scrum/models/project_task.py:
```diff
@@ -0,0 +1,124 @@
+        return [('sprint_id', '=', sprint and sprint.id or None)]
+
+    @api.model
+    def _read_group_sprint_id(self, present_ids, domain, **kwargs):
```
are you missing `group_expand` clauses for the involved fields here and below?

hbrunn commented on project_scrum/views/project_project.xml:
```diff
@@ -0,0 +1,160 @@
+        <field
+            name="search_view_id"
+            ref="project.view_project_project_filter" />
+        <field name="context">{'default_use_scrum': True}</field>
```
you need to force using the scrum specific views here and in all actions below

hbrunn commented on project_scrum/views/project_project.xml:
```diff
@@ -0,0 +1,165 @@
+        <field
+            name="search_view_id"
+            ref="project.view_project_project_filter" />
+        <field name="context">{'default_use_scrum': True}</field>
```
there should be a domain on this one to actually only show scrum projects. I think this is my last nag

daramousk replied:
@hbrunn Implemented. (not sure if you are getting notifications about my pushes)

hbrunn replied:
thank $deity not!
